modules/core/src/smithy4s/http/HttpErrorSelector.scala

```diff
@@ -80,20 +79,24 @@ private[http] final class HttpErrorSelector[F[_]: Covariant, E](
   type ConstF[A] = F[E]
   val cachedDecoders: PolyFunction[Alt[E, *], ConstF] =
     new PolyFunction[Alt[E, *], ConstF] {
-      def apply[A](alt: Alt[E, A]): F[E] = {
+      def compileAlt[A](alt: Alt[E, A]): F[E] = {
         val schema = alt.schema
-        // TODO: apply proper memoization of error instances/
         // In the line below, we create a new, ephemeral cache for the dynamic recompilation of the error schema.
         // This is because the "compile entity encoder" method can trigger a transformation of hints, which
         // lead to cache-miss and would lead to new entries in existing cache, effectively leading to a memory leak.
         val cache = compiler.createCache()
         val errorCodec: F[A] = compiler.fromSchema(schema, cache)
         Covariant[F].map[A, E](errorCodec)(alt.inject)
       }
-    }.unsafeCacheBy(
-      alts.map(Kind1.existential(_)),
-      identity(_)
-    )
+      val builder = Map.newBuilder[Any, Any]
+      alts.foreach { alt =>
+        builder += alt -> compileAlt(alt)
+      }
+      val resultCache = builder.result()
+      def apply[A](alt: Alt[E, A]): F[E] = {
+        resultCache(alt).asInstanceOf[F[E]]
```

Comment:
**question:** is this something we could encapsulate in a public pattern? I suppose it'd be a sort of dual to `Dispatcher`

Reply:
We might be able to but I've not thought about it enough to know if it's possible

Reply:
Right, the bare minimum for us to be able to generalise would be to have a flatmap constraint on the F

Reply:
I'd be fine with such a constraint. `flatten` + `map` (from Covariant) would be fine, no?

Reply:
I gave it a go and it's a lot more complicated to generalise than the encoding side, partially because the action of decoding may "consume" data. 

So it ends up looking like this : 

```scala
    def compileDecoder[Decoder[_], F[_], Message, Discriminator](
        precompile: Precompiler[Decoder],
        discriminator: Alt[_, _] => Discriminator,
        discriminate: Message => F[(Discriminator, Message)]
    )(implicit
        decoder: DecoderK[Decoder, F, Message]
    ): Decoder[U]
```

It may be worth exploring further but certainly not in this PR